### PR TITLE
ClassVariable with #definingClass and related cleanup and refactoring 

### PR DESCRIPTION
**Pull request by dionisiydk**
To follow Slots design and use cases PR adds #definingClass, #owningClass and related methods into ClassVariable. 
There is little duplication with Slot code. But I don't think we can address it easily because ClassVariable requires key->value object layout (see #pushLiteralVariable: ). 

Other changes are related to this unification between Slot and ClassVariable API:
- All special class variable classes are now subclasses of ClassVariable (WeakClassVariable and others).
- Slot>>#isLocal is deprecated by #isDefinedByOwningClass to not overlap with temp variables #isLocal meaning. ClassVariable also defines #isDefinedByOwningClass and we now can support traits.
- Dictionary>>declareVariable:from: is removed. It is replaced by Class method #basicDeclareClassVariable:. I discovered exiting issues here and put a note into comment and todo flag (I will add an issues later)
- little change in FFIEnumeration code which builds class variables (@estebanlm please review). It did not follow class addition API which I think was wrong.

### CODE REVIEW

**Comment by dionisiydk** [comment]
More work is needed

**Review by MarcusDenker — APPROVED** [review]
Ok

**Comment by dionisiydk** [comment]
It needs more work after Pablo fix for shared vars

вт, 7 июл. 2020 г. в 11:06, Marcus Denker <notifications@github.com>:

**Comment by MarcusDenker** [comment]
Ok, I will make sure to merge Pablo's change first

File: bootstrap/src/Pharo30Bootstrap/PBClassLoader.class.st [hunk 1]
```diff
@@ -207,9 +207,12 @@ PBClassLoader >> initializeClassPool: aClassName [
 
 	| definition class classPool classVariableBindings |
 	definition := systemDefinition classNamed: aClassName.
+	class := environment at: aClassName.
+
 	classVariableBindings := definition classVarNames collect: [ :each | | classVar |
 		classVar := (self classNamed: #ClassVariable) basicNew.
-		classVar instanceVariableAtIndex: 1 put: (each asLiteralInObjectSpace: builder objectSpace) ].
+		classVar instanceVariableAtIndex: 1 put: (each asLiteralInObjectSpace: builder objectSpace).
+		classVar instanceVariableAtIndex: 3 put: class "initialize owningClass" ].
```
Comment (dionisiydk) [inline bootstrap/src/Pharo30Bootstrap/PBClassLoader.class.st:215]: @guillep could you review this bootstrap change please? Tests shows that it's fine. But maybe it should be done differently. 
Anyway I do not think it prevents merging. I can address your suggestion (if any) in another PR

**Comment by dionisiydk** [comment]
It is now ready for merge

**Pull request merged**